proposals/0000-no-fallible-do.rst

```diff
@@ -0,0 +1,112 @@
+            number in the link, and delete this bold sentence.**
+.. contents::
+
+Motivation
+----------
+There are situations in which pattern completeness checking and ``do``-syntax can presently interact in frustrating ways. On the one hand, there are cases involving GADTs where one knows the index type of a GADT being pattern matched, and said GADT has only a single constructor with that index, making the pattern match theoretically complete, but GHC presently fails to discover this. For example:
+
+::
+
+  data MyGADT :: * -> * where
+    A :: Int -> MyGADT Int
+    B :: MyGADT Bool
+  
+  example :: (Monad m) => m (MyGADT Int) -> m Int
+  example x = do
+    A n <- x
+    return n
+
+This presently fails because the desugaring of the ``do``-expression involves a use of ``fail``, which induces a ``MonadFail`` constraint, even though the generated ``fail`` is theoretically dead code.
+
+There is a bit of a Gordian knot in this, in that the type of a ``do``-expression straightforwardly depends on how it is desugared, but if we improve the completeness checker and make it operate in time to handle this, the syntactic meaning of a ``do``-expression may then depend on its type as well. In non-contrived scenarios, I don't expect that apparent dependency cycle to be a problem, but it still complicates things from the perspective of determining where in the compiler this should all take place.
+
+It's also unclear in general whether we want the unfolding of syntax sugar to depend on anything that is going on at the type level in the first place, as that makes it far more challenging to explain and understand.
+
+On the other hand, there are additionally cases where completeness checking of pattern synonyms becomes quite challenging as well, and ``{-# COMPLETE #-}`` pragmas may presently be required to convince GHC that certain pattern matches are complete. Even if those are specified, at present, the information about which ``COMPLETE`` pragmas exist is not available at the time of desugaring ``do``-syntax. Both making GHC take ``COMPLETE`` pragmas into consideration when deciding how to desugar `do`-expressions, and making further improvements to completeness checking in the face of pattern synonyms seem like good ideas, but there's certainly more work to be done in sorting out how this will actually be achieved in the context of GHC.
+
+These problems are frustrating in context, because often one has no interest in having ``MonadFail`` constraints appear at all, so one might want to avoid going down a rabbit hole of subtler issues of completeness checking when the real concern is for ``fail`` not to appear in the translation at all.
+
+So to cut through the knots a bit and provide a way to avoid the more complicated and challenging issues about completeness checking while their solutions are worked out, we decided to try something simple and provide a way to turn off the use of ``fail`` altogether.
```

Comment:
```suggestion
These problems are frustrating in context, because often one has no interest in having ``MonadFail`` constraints appear at all.
Errors made with `fail` are always strings, and so completely unstrucutred. Whether one prefers `Either` and `EitherT` or synchronous exceptions, the norm is to use types to structure failure modes, just as we use types to structure everything else in Haskell.
Furthermore, even if one does want to use unstructured textual errors, `fail` uses `String` rather than something more performant like `Text`.
This is cumbersome when the rest of the program uses `Text`.
For these reasons, one might want to avoid going down a rabbit hole of subtler issues of completeness checking when the real concern is for ``fail`` not to appear in the translation at all.

So to cut through the knots a bit and provide a way to avoid the more complicated and challenging issues about completeness checking while their solutions are worked out, we decided to try something simple and provide a way to turn off the use of ``fail`` altogether.
```